protocol/messenger.go

```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
+	if err := m.SubscribeToPubsubTopic(shard.DefaultNonProtectedPubsubTopic(), nil); err != nil {
+		return err
+	}
+
+	filters, publicKeys, err := m.collectFiltersAndKeys()
 	if err != nil {
 		return err
 	}
 
-	var (
-		filtersToInit []transport.FiltersToInitialize
-		publicKeys    []*ecdsa.PublicKey
-	)
+	_, err = m.transport.InitFilters(filters, publicKeys)
+	return err
+}
+
+func (m *Messenger) collectFiltersAndKeys() ([]transport.FiltersToInitialize, []*ecdsa.PublicKey, error) {
+	var wg sync.WaitGroup
+	errCh := make(chan error, 5)
+	filtersCh := make(chan []transport.FiltersToInitialize, 3)
+	publicKeysCh := make(chan []*ecdsa.PublicKey, 2)
+
+	wg.Add(5)
+	go m.processJoinedCommunities(&wg, filtersCh, errCh)
+	go m.processSpectatedCommunities(&wg, filtersCh, errCh)
+	go m.processChats(&wg, filtersCh, publicKeysCh, errCh)
+	go m.processContacts(&wg, publicKeysCh, errCh)
+	go m.processControlledCommunities(&wg, errCh)
+
+	wg.Wait()
+	close(filtersCh)
+	close(publicKeysCh)
+
+	select {
+	case err := <-errCh:
+		return nil, nil, err
+	default:
+	}
+
+	return m.collectResults(filtersCh, publicKeysCh)
+}
+
+func (m *Messenger) processJoinedCommunities(wg *sync.WaitGroup, filtersCh chan<- []transport.FiltersToInitialize, errCh chan<- error) {
+	defer gocommon.LogOnPanic()
+	defer wg.Done()
 
 	joinedCommunities, err := m.communitiesManager.Joined()
 	if err != nil {
-		return err
+		errCh <- err
+		return
 	}
-	for _, org := range joinedCommunities {
+
+	filtersToInit := m.processCommunitiesSettings(joinedCommunities)
+	filtersCh <- filtersToInit
+}
+
+func (m *Messenger) processCommunitiesSettings(communities []*communities.Community) []transport.FiltersToInitialize {
+	logger := m.logger.With(zap.String("site", "processCommunitiesSettings"))
+	var filtersToInit []transport.FiltersToInitialize
+
+	for _, org := range communities {
 		// the org advertise on the public topic derived by the pk
 		filtersToInit = append(filtersToInit, m.DefaultFilters(org)...)
 
-		// This is for status-go versions that didn't have `CommunitySettings`
-		// We need to ensure communities that existed before community settings
-		// were introduced will have community settings as well
-		exists, err := m.communitiesManager.CommunitySettingsExist(org.ID())
-		if err != nil {
-			logger.Warn("failed to check if community settings exist", zap.Error(err))
-			continue
+		if err := m.ensureCommunitySettings(org); err != nil {
+			logger.Warn("failed to process community settings", zap.Error(err))
 		}
+	}
 
-		if !exists {
-			communitySettings := communities.CommunitySettings{
-				CommunityID:                  org.IDString(),
-				HistoryArchiveSupportEnabled: true,
-			}
+	return filtersToInit
+}
 
-			err = m.communitiesManager.SaveCommunitySettings(communitySettings)
-			if err != nil {
-				logger.Warn("failed to save community settings", zap.Error(err))
-			}
-			continue
-		}
+func (m *Messenger) ensureCommunitySettings(org *communities.Community) error {
```

Comment:
Actually it looks like this function doesn't really depend on the `Messenger`, only the `CommunityManager`. Would be even better to move it there.